### ISSUE TITLE
Quality control of dates and titles in records and motions

### ISSUE THREAD

**Issue by MansMeg** [swerik-project/the-swedish-parliament-corpus#44]
The library is interested in assessing the quality of the dates (motions and records) and the titles (motions).

The date of interest is the start and end date of each record and the date of submission of the motion. The title of interest is the title of the motions. 
The goal from RB is 98% accuracy

To assess the quality we need to do the following:
- [ ] define a sampling plan and estimation plan and add to the documentation as a quality dimension(we take a random sample of 3 records per year/and document and manually check the dates)
- [ ] do manual annotations
- [ ] implement estimation

**Comment by MansMeg** [comment]
Now as PR to motions and records.

**Comment by BobBorges** [comment]
Samples for manual annotation are here:
---
Motions: dates and titles ([here](https://docs.google.com/spreadsheets/d/19srg5BJaKYbZvE6LwX0-L-bGJxvpj64Uv0_QB3FbQD8/edit?gid=0#gid=0))

Records: dates ([here](https://docs.google.com/spreadsheets/d/1wMY_mqsxltZoPV8Zr5FO4Nt8kq0yYC9mTMXUJMY3V-Q/edit?gid=0#gid=0))


@fredrik1984 @Lottabrorsson, You should already have edit rights on these docs, but let me know if not.

**Comment by fredrik1984** [comment]
How is it now, what should we write before, e.g., "riksdagen-motions/data/1867/mot-1867--ak--00233.xml" to access the document?

**Comment by BobBorges** [comment]
Sorry -- I put full URLs in the sheets now.

**Comment by fredrik1984** [comment]
find the csv with date quality for the records and add that to the quality folder in the repo.

**Comment by ninpnin** [comment]
@fredrik1984 AFAIK, the file corresponding to "Here's the corrected gold standard" is what we want to add to the repo https://github.com/swerik-project/riksdagen-records/issues/16

**Comment by MansMeg** [comment]
We are waiting for annotations from @Lottabrorsson and @fredrik1984 .

**Comment by fredrik1984** [comment]
Have I not done my share of this? Is it Lotta's share you are waiting for?

**Comment by MansMeg** [comment]
Fredrik has done his part. Waiting for Lotta.

**Comment by MansMeg** [comment]
Waiting for Lotta on the records. Bob will start to work with the motions now.

**Comment by MansMeg** [comment]
Code and results for motions, hopefully done today by Bob.

**Comment by BobBorges** [comment]
related https://github.com/swerik-project/riksdagen-records/pull/107

**Comment by MansMeg** [comment]
Adding data done by Monday by @mandlilaast .

**Comment by MansMeg** [comment]
Will be done when PR #107 is merged.

**Comment by mandlilaast** [comment]
Hi!  

The PR regarding this issue is waiting for approval in **PR #125** in *riksdagen-records*.  

Sorry that it took longer than expected — the gold-standard student data had not been included, and the code was not yet implemented (in one of the branches, the file existed with a comment: *“to be implemented :D”*).  

Additionally, the structure of the current student data (using `to` and `from`, but actual protocols sometimes span 4–5 days) required manual corrections, as the automatic approach failed. This changed the results significantly:  

- **Before fixes:** 201 missing dates, 79 wrong annotations  
- **After fixes:** 6 missing dates, 373 (possibly) wrong annotations  

Unfortunately, due to the nature of the problem, I cannot give accuracy per se, but rather **precision** and **recall** metrics:  

- **Missing annotation:** considered as false negative  
- **Existing but wrong annotation:** considered as false positive  

Metrics:  

| Dataset   | Precision | Recall  | F1     |
|-----------|-----------|---------|--------|
| Expert    | 83.36%    | 99.86%  | 90.87% |
| Student   | 82.77%    | 99.50%  | 90.37% |
| Combined  | 83.23%    | 99.68%  | 90.71% |

While going over the false negatives, I noticed the following discrepancies:  

1. **Expert data:**  
   - `1924--ak--018.xml` ([PDF](https://pdf.swedeb.se/riksdagen-records-pdf/1924/prot-1924--ak--018.pdf)) — listed as `ak`, but path directs to `fk`. Same issue with the next link ([PDF](https://pdf.swedeb.se/riksdagen-records-pdf/1924/prot-1924--ak--019.pdf)).  

2. **Student data:**  
   - `prot-1888--ak--023.xml` ([PDF](https://pdf.swedeb.se/riksdagen-records-pdf/1888/prot-1888--ak--023.pdf)) — protocol placed in the wrong folder (should be 1887). Bob notified Lotta about it recently.  

Other false negatives were rather logical:  

- `prot-1879--ak--059.xml` — OCR reads `1` as `I`, and therefore the automatic approach does not recognize it as date.
- `prot-199293--127.xml` — OCR breaks title due to two columns in the XML; date spread across parts, not recognized  

Maybe a possibility would be to continue with a new issue regarding the **false positives**, if necessary. 🙂

**Comment by MansMeg** [comment]
Hi! This is great! 

Im not sure I follow perfectly, but the small issues you identfied, like 1 and 2, would be great to file in a new issue. Then we will talk with Lotta and the Library on what the level they want.

- [ ] Open new issue with small errors to fix.

Also, Im not sure why we cannot compute accuracy?

**Comment by mandlilaast** [comment]
@MansMeg I hope that I’m not mistaken, but since we have the gold standard, we only work with detected dates — meaning that the **True Positive (TP)**, **False Positive (FP)**, and **False Negative (FN)** values can be found. 

From what I understand, we cannot really calculate accuracy in this setup, since it would require the fourth metric, **True Negative (TN)**:

$$
\text{Accuracy} = \frac{TP + TN}{TP + TN + FP + FN}.
$$

The **TN** here would, to my understanding, correspond to every non-date segment that was correctly *not* classified as a date. Due to the huge number of such non-date segments relative to those classified in any way, the accuracy would be approximately 1, and therefore would not reflect the actual performance of date detection.

The recall (or true positive rate) is a better metric here, since it tells us how many of the real document dates were correctly identified. Together with precision, it shows the real trade-off: the current date detection algorithm seems quite sensitive, often classifying references or contextual dates as document dates.

**Comment by MansMeg** [comment]
Yes. I dont remember the sampling process. But dont we just sample record and then check the dates of that record? Ie its not a classification problem but rather how many days off/jaccard?

**Comment by ninpnin** [comment]
I think the accuracy would be calculated as the proportion of the documents with the correct date range / date set. The labeling of individual paragraphs is less important.

For each estimated date range $\hat{D}_i$ and true date range $D_i$, the Jaccard coefficient would be calculated as the 

$$
J_i = J(\hat{D}_i, B) = \frac{|\hat{D}_i \cap D|}{|\hat{D}_i \cup D|}
$$

This can be then averaged to get $\mathbb{E}[J_i]$ (avg. Jaccard), $P(J_i = 1)$ (accuracy), $P(D \subset \hat{D}_i)$ etc.

**Comment by MansMeg** [comment]
This is done in #127
